### DISCUSSION TITLE
Feature Request: compression level option for EXPORT command

### DISCUSSION THREAD

**Discussion by traversc** [duckdb/duckdb#10170] [Ideas]
I have a database that I'm trying to optimize for size

```
con.sql("EXPORT DATABASE 'my_db' (FORMAT PARQUET, COMPRESSION ZSTD)")
```

The resulting size is about 50% larger than the raw input parquet (which was compressed with maximum zstd compression level). It would be a nice option if you could increase the compression level.

**Comment by marklit** [comment]
I'd love to see this feature in DuckDB.

I patched the source to see what sort of effect it would have on an existing PQ file produced in DuckDB with level 3 compression (which is ZSTDs default).

```diff
diff --git a/third_party/zstd/include/zstd.h b/third_party/zstd/include/zstd.h
index ade94c2d5c..80d4b560a9 100644
--- a/third_party/zstd/include/zstd.h
+++ b/third_party/zstd/include/zstd.h
@@ -84,7 +84,7 @@ ZSTDLIB_API const char* ZSTD_versionString(void);   /* requires v1.3.0+ */
  *  Default constant
  ***************************************/
 #ifndef ZSTD_CLEVEL_DEFAULT
-#  define ZSTD_CLEVEL_DEFAULT 3
+#  define ZSTD_CLEVEL_DEFAULT 22
 #endif

 /* *************************************
```

I managed to reduce an example Parquet file by 0.83x. DuckDB effectively processed the PQ file at 4.55 MB/s and 104K records/s. 338,609 records and a very wide number of fields compressed at level 22 in 3.251 seconds. This was on an Intel Core i9-14900K. When compressing a deliverable for a wide audience or archiving the compression rate shouldn't matter so much but still, it isn't prohibitively slow.

```sql
INSTALL '/mnt/d/duckdb_source/build/release/extension/parquet/parquet.duckdb_extension';
LOAD parquet;

COPY(
    SELECT *
    FROM READ_PARQUET('/mnt/f/gis/Global/adsb_lol/2024/traces_2024-02-29_0045.pq')
    ORDER BY trace.lon
) TO 'test.zstd.22.pq' (FORMAT 'PARQUET',
                        CODEC  'ZSTD',
                        ROW_GROUP_SIZE 15000);
```

Here are a few other compression levels for comparison.

Level | Bytes
------|------
 3 | 15,523,193
 9 | 14,459,408
17 | 13,117,006
22 | 12,888,833

**Comment by BGameiro2000** [comment]
This would be really helpful.

Perhaps it could be added as configuration that can be set (`SET ZSTD_CLEVEL TO 22;`)?
The same way we can set `SET threads TO 4;`

**Comment by Mytherin** [comment]
Implemented in https://github.com/duckdb/duckdb/pull/11791
